### ISSUE TITLE
Android: choosing a different controller preset in dosbox-pure config crashes Retroarch 

### ISSUE THREAD

**Issue by svabos** [schellingb/dosbox-pure#511]
I'm experiencing constant crash on andorid even after pure reinstall. Retroarch version: 1.19.1 ; Dosbox-pure version: 0.9.9  

When loading a dosbox-pure core and going into it's menu "controller mapper", then trying to change preset from default "generic keyboard" the retroarch just crashes. 
That doesn't happen on PC version of retroarch/dosbox-pure

Bonus:
is there a way to map a gyroscope as mouse movement in core controller mapper

**Comment by schellingb** [comment]
I think this was fixed in 1c7ad4f. It should get released soon, I don't know when yet though. Some big changes since the last release.

Controlling something with a gyroscope sounds like a neat idea, do you know any other core that supports that? I'm not sure if it's available at all to a libretro core...

**Comment by svabos** [comment]
> Controlling something with a gyroscope sounds like a neat idea, do you know any other core that supports that? I'm not sure if it's available at all to a libretro core...

yes, a mGBA core has it and works on android phone. I just tested it and can confirm that it works with game "WarioWare: Twisted!"
Other than that, gyro control is highly requested feature on numerous cores but I haven't seen it anywhere else

**Comment by schellingb** [comment]
Ok, looks like there's this part in the libretro code interface

```c
/* Id values for SENSOR types. */
#define RETRO_SENSOR_ACCELEROMETER_X 0
#define RETRO_SENSOR_ACCELEROMETER_Y 1
#define RETRO_SENSOR_ACCELEROMETER_Z 2
#define RETRO_SENSOR_GYROSCOPE_X 3
#define RETRO_SENSOR_GYROSCOPE_Y 4
#define RETRO_SENSOR_GYROSCOPE_Z 5
#define RETRO_SENSOR_ILLUMINANCE 6

typedef bool (RETRO_CALLCONV *retro_set_sensor_state_t)(unsigned port,
      enum retro_sensor_action action, unsigned rate);

typedef float (RETRO_CALLCONV *retro_sensor_get_input_t)(unsigned port, unsigned id);

struct retro_sensor_interface
{
   retro_set_sensor_state_t set_sensor_state;
   retro_sensor_get_input_t get_sensor_input;
};
```

It wouldn't be easy to add support for it in DOSBox Pure. But maybe some day :-)

**Comment by svabos** [comment]
> Ok, looks like there's this part in the libretro code interface
> 
> ```c
> /* Id values for SENSOR types. */
> #define RETRO_SENSOR_ACCELEROMETER_X 0
> #define RETRO_SENSOR_ACCELEROMETER_Y 1
> #define RETRO_SENSOR_ACCELEROMETER_Z 2
> #define RETRO_SENSOR_GYROSCOPE_X 3
> #define RETRO_SENSOR_GYROSCOPE_Y 4
> #define RETRO_SENSOR_GYROSCOPE_Z 5
> #define RETRO_SENSOR_ILLUMINANCE 6
> 
> typedef bool (RETRO_CALLCONV *retro_set_sensor_state_t)(unsigned port,
>       enum retro_sensor_action action, unsigned rate);
> 
> typedef float (RETRO_CALLCONV *retro_sensor_get_input_t)(unsigned port, unsigned id);
> 
> struct retro_sensor_interface
> {
>    retro_set_sensor_state_t set_sensor_state;
>    retro_sensor_get_input_t get_sensor_input;
> };
> ```
> 
> It wouldn't be easy to add support for it in DOSBox Pure. But maybe some day :-)

Please do add it! It's the only hope for playing mouse games on andorid. Touchscreen is just pain to use...

**Comment by schellingb** [comment]
Problem I can see is that it would need a lot of special options.
There are 6 axis available (gyro XYZ and accel XYZ). How to set which one does what? How to set speed? Would it be enough to just control the mouse? Or would one also want to control other inputs. I.e. joysticks. Or generic keys...

**Comment by svabos** [comment]
I'm no expert, but "gyro" sets the direction and "accel" sets the speed/sensitivity. We don't need Z axis emulation...
Just the mouse would be more that sufficient for now

I understand that all of the gyro inputs could be used, but as cool as it could be, only tilting X and Y axes it more that enough for games that require mouse, we don't need to invent "hot water" here ;)

**Comment by tails101** [comment]
I wonder if it could be used for a wheel ,for cars or maybe flight simulator.
